sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala

```diff
@@ -1036,8 +1036,17 @@ case class SortMergeJoinExec(
     val rightResultVars = genOneSideJoinVars(
       ctx, rightOutputRow, right, setDefaultValue = true)
     val resultVars = leftResultVars ++ rightResultVars
-    val (_, conditionCheck, _) =
-      getJoinCondition(ctx, leftResultVars, left, right, Some(rightOutputRow))
+    // Evaluate the variables on the left and used in the condition but do not clear the code as
```

Comment:
This is very hard to review. Is there any similar code in other join types or `ShuffledHashJoinExec`? It seems other places are using `splitVarsByCondition` and `evaluateVariables`.

Reply:
Thanks @cloud-fan for your comment.
 
1. `ShuffledHashJoinExec` also has the same issue, and fix it at the same time.
2. We can not use `splitVarsByCondition` and `evaluateVariables`  because `evaluateVariables` will empty the variable's code, so when we read the same variables in the following `consumeFullOuterJoinRow` method, those variables will be undefined.

Reply:
> because evaluateVariables will empty the variable's code

can we copy the variables and invoke `evaluateVariables` with the copies? Then the original variables are unchanged and can be safely accessed later.

Reply:
Yes, you are right, I updated the code and it is much clearer than before. 
Thanks for your help.